Core/include/Acts/Detector/ComponentBuilderProxy.hpp

```diff
@@ -0,0 +1,168 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2023 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/Definitions/Algebra.hpp"
+#include "Acts/Geometry/VolumeBounds.hpp"
+#include "Acts/Utilities/BinningData.hpp"
+
+#include <memory>
+#include <string>
+
+namespace Acts {
+namespace Experimental {
+
+class IDetectorComponentBuilder;
+
+/// @brief A container proxy struct that can be used to build up
+/// a tree graph of detector components
+class ComponentBuilderProxy
+    : public std::enable_shared_from_this<ComponentBuilderProxy> {
+ protected:
+  /// Createing a ComponentBuilderProxy is not allowed
+  /// @note use : createRootProxy(), createChildVolumeProxy() or
+  /// createChildContainerProxy()
+  ComponentBuilderProxy() = default;
+
+ public:
+  /// @brief Private constructor to be used for the proxy creation
+  /// @param n The name of the proxy
+  ComponentBuilderProxy(const std::string& n) : name(n){};
+
+  /// @brief  Definition of a volume
+  struct VolumeProxy {
+    /// If fully defined, the builder is set
+    std::shared_ptr<const IDetectorComponentBuilder> builder = nullptr;
+  };
+
+  /// @brief  Definition of a container
+  struct ContainerProxy {
+    /// The binning instruction
+    std::vector<BinningValue> binning = {};
+    /// The contained component builders, in case of a container builder
+    std::vector<std::shared_ptr<ComponentBuilderProxy>> children = {};
+  };
+
+  /// The proxy name - this needs to be uniuque
+  std::string name = "";
+
+  /// The Parent proxy, optionally set
+  std::shared_ptr<ComponentBuilderProxy> parent = nullptr;
+
+  /// The volume proxy, optionally set
+  std::variant<VolumeProxy, ContainerProxy> holder = VolumeProxy{};
```

Comment:
If I'm reading correctly what you're doing with this, wouldn't this be much easier done using another interface and inheriting from it? Then you can get rid of the manual checking of alternatives, and just use virtual function calls, maybe.

Reply:
I would have to do dynamic_cast then, I don't mind - either way works.

Reply:
But can't you have a method which just returns what you create in the `if`/`else` now?